src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -574,21 +583,30 @@ private static Task GetHostEntryOrAddressesCoreAsync(string hostName, bool justR
             }
         }
 
-        private static async Task<T> GetAddrInfoWithTelemetryAsync<T>(string hostName, bool justAddresses, AddressFamily addressFamily, CancellationToken cancellationToken)
-            where T : class
+        private static  bool TryGetAddrInfoWithTelemetryAsync<T>(string hostName, bool justAddresses, AddressFamily addressFamily, CancellationToken cancellationToken, out Task t)
+             where T : class
         {
             ValueStopwatch stopwatch = NameResolutionTelemetry.Log.BeforeResolution(hostName);
 
-            T? result = null;
             try
             {
-                result = await ((Task<T>)NameResolutionPal.GetAddrInfoAsync(hostName, justAddresses, addressFamily, cancellationToken)).ConfigureAwait(false);
-                return result;
+                if (NameResolutionPal.TryGetAddrInfoAsync(hostName, justAddresses, addressFamily, cancellationToken, out t))
+                {
+                    t.ContinueWith((resultTask) => {
```

Comment:
We should be returning the Task returned by `ContinueWith`, or the ActivityId following the Dns resolution wouldn't reflect the Stop event.
This was the issue with the original Telemetry instrumentation of Dns, pointed out by @ManickaP in https://github.com/dotnet/runtime/issues/41670#issuecomment-685038205 (and fixed in #42188).

Reply:
Yep, this will break the telemetry once again.

Reply:
I had different change where telemetry test was failing. With this on, telemetry tests are still passing. The log still be edited but I'm not sire about the ID. If we return the ContinueWith, we will need to copy result (or failure). I will look into it.

Reply:
The tests will only validate that the Start/Stop are logged with the same ActivityId.

The issue here appears when you have another Start/Stop pair following the name resolution:
- HttpRequestStart starts with ID `/1/`
- NameResolutionStart starts with nested ID `/1/1/`
- NameResolutionStop correctly sees `/1/1/` and updates it to `/1/`.
- SocketConnectStart starts on the same async context as `NameResolutionStart`, but it never observes the change of the ID back to `/1/`, so it starts with a nested ID `/1/1/1`

It will appear as if the Socket connect is an operation nested under the NameResolution Start/Stop, when in reality it's only following it.

Reply:
I re-factor the code and the activity ID is right as far as I can tell as well as running impersonificated still works.

I went back to the `await` path as the `ContinueWith` was getting complicated more than I wanted. The only downside is that we will miss time of the call it self. 
I feel it probably does not matter comparing duration of the actual resolution. 
We could probably start the timer upfront but I'm not sure if that is worth of the complication.

Reply:
after chatting with @MihaZupan , I decided to fix the stopwatch as well. 
This should be ready for another review.

Reply:
Telemetry here LGTM